jlm/util/BiMap.hpp

```diff
@@ -0,0 +1,254 @@
+    if (it != ForwardMap_.end())
+      throw jlm::util::error("Key already exsists in BiMap");
+    AddPairToReverse(key, value);
+    ForwardMap_.insert({ key, value });
+  }
+
+  /**
+   * Checks for the existance of the given \p key in the BiMap
+   * @param key the key to look for
+   * @returns true if the key exists in the BiMap, otherwise false.
+   */
+  [[nodiscard]] bool
+  HasKey(const K & key) const noexcept
+  {
+    return ForwardMap_.find(key) != ForwardMap_.end();
+  }
+
+  /**
+   * Checks for the existance of the given \p value in the BiMap
+   * @param value the value to look for
+   * @returns true if the value exists in the BiMap, otherwise false.
+   */
+  [[nodiscard]] bool
+  HasValue(const V & value) const noexcept
+  {
+    return ReverseMap_.find(value) != ReverseMap_.end();
+  }
+
+  /**
+   * @param key the key to lookup in the BiMap
+   * @returns the value associated with the given \p key
+   * @throws jlm::util::error if the \p key does not exist in the BiMap
+   */
+  [[nodiscard]] const V &
+  LookupKey(const K & key) const
+  {
+    auto it = ForwardMap_.find(key);
+    if (it == ForwardMap_.end())
+      throw jlm::util::error("Key not found in BiMap");
+    return it->second;
+  }
+
+  /**
+   * @param value the value to lookup in the BiMap
+   * @returns the key associated with the given \p value
+   * @throws jlm::util::error if the \p value does not exist in the BiMap
+   */
+  [[nodiscard]] const K &
+  LookupValue(const V & value) const
+  {
+    auto it = ReverseMap_.find(value);
+    if (it == ReverseMap_.end())
+      throw jlm::util::error("Value not found in BiMap");
+    return it->second;
+  }
+
+  /**
+   * Removes a key/value pair from the BiMap, if the given \p key exists
+   * @param key the key to be removed
+   * @returns true if a key/value pair was removed
+   */
+  bool
+  RemoveKey(const K & key)
+  {
+    auto it = ForwardMap_.find(key);
+    if (it == ForwardMap_.end())
+      return false;
+
+    auto removed = ReverseMap_.erase(it->second);
+    JLM_ASSERT(removed == 1);
+
+    ForwardMap_.erase(it);
+    return true;
+  }
+
+  /**
+   * Removes a key/value pair from the BiMap, if the given \p value exists
+   * @param value the value to be removed
+   * @returns true if a key/value pair was removed
+   */
+  bool
+  RemoveValue(const V & value)
+  {
+    auto it = ReverseMap_.find(value);
+    if (it == ReverseMap_.end())
+      return false;
+
+    auto removed = ForwardMap_.erase(it->second);
+    JLM_ASSERT(removed == 1);
+
+    ReverseMap_.erase(it);
+    return true;
+  }
+
+  /**
+   * @returns the underlying mapping from keys to values
+   */
+  [[nodiscard]] const std::unordered_map<K, V> &
+  GetForwardMap() const noexcept
+  {
+    return ForwardMap_;
+  }
+
+  /**
+   * @returns the underlying mapping from values to keys
+   */
+  [[nodiscard]] const std::unordered_map<V, K> &
+  GetReverseMap() const noexcept
+  {
+    return ReverseMap_;
+  }
```

Comment:
I really do not like these two functions as they are basically exposing the internal implementation of the bijective map to the caller. Why do you need them?